File: solution39.java


import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class solution39 {
    private List<List<Integer>> res;

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        res = new ArrayList<>();
        dfs(candidates, new LinkedList<Integer>(), target, 0);
        return res;
    }

    public void dfs(int[] candidates, LinkedList<Integer> path, int target, int startIndex) {
        if (startIndex == candidates.length) {
            return;
        }
        if (target == 0) {
            res.add(new ArrayList<>(path));
            return;
        }
        if (target < 0) {
            return;
        }
        if (candidates[startIndex] <= target) {
            path.add(candidates[startIndex]);
            dfs(candidates, path, target - candidates[startIndex], startIndex);
            path.removeLast();
        }
        dfs(candidates, path, target, startIndex + 1);
    }
}

File: Cxx/solution39.cpp
#include <vector>
using namespace std;

class Solution {
    void dfs(vector<int> &candidates, vector<int> &temp, int index, int target,
             vector<vector<int> > &res) {
        if (index == candidates.size()) {
            return;
        }
        if (target == 0) {
            res.push_back(temp);
            return;
        }
        if (target - candidates[index] >= 0) {
            temp.push_back(candidates[index]);
            dfs(candidates, temp, index, target - candidates[index], res);
            temp.pop_back();
        }
        dfs(candidates, temp, index + 1, target, res);
    }

   public:
    vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
        vector<vector<int> > res;
        vector<int> temp;
        dfs(candidates, temp, 0, target, res);
        return res;
    }
};